### PR TITLE
feat(experiments): add on_error callback to flush and experiment run

### PR DESCRIPTION
**Pull request by thiagobomfin-galileo**
# User description
**Shortcut:**
[SC-56022](https://app.shortcut.com/galileo/story/56022/run-experiment-hides-flush-errors-causing-0-trace-failures)

**Description:**

Adds an optional on_error parameter to GalileoDecorator.flush(), Experiments.run(), Experiments.run_with_function(), and run_experiment().
When provided, flush errors log at debug level and invoke the callback instead of logging a warning. Experiments.run() also uses on_error to handle creation failures without raising when a callback is supplied.

**Tests:**

- [x] Unit Tests Added
- [ ] [E2E Test](https://github.com/rungalileo/e2e-testing) Added (if it's a user-facing feature, or fixing a bug)

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
Add optional <code>on_error</code> callbacks to the flush pipeline so callers can handle ingestion failures without flooding warnings from <code>GalileoDecorator</code> and <code>GalileoLogger</code>. Propagate the callback through the <code>Experiments</code> flows and <code>run_experiment</code> to keep creation failures raised while warning when the hook is unused in the prompt-template path.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/rungalileo/galileo-python/537?tool=ast&topic=Flush+error+handling>Flush error handling</a>
        </td><td>Guard flush errors by logging at debug when a callback exists, invoking the user-supplied <code>on_error</code>, and warning if the callback itself raises.<details><summary>Modified files (3)</summary><ul><li>src/galileo/decorator.py</li>
<li>src/galileo/logger/logger.py</li>
<li>tests/test_decorator.py</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>fernando.correia@galil...</td><td>chore: update ruff tar...</td><td>April 07, 2026</td></tr>
<tr><td>pratyusha@galileo.ai</td><td>feat: serialize trace ...</td><td>March 24, 2026</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/rungalileo/galileo-python/537?tool=ast&topic=Experiment+on_error+path>Experiment on_error path</a>
        </td><td>Extend the experiment flows to accept <code>on_error</code>, route it into the function-run flushes, and warn when it cannot be used in the prompt-template path while keeping creation failures raised.<details><summary>Modified files (2)</summary><ul><li>src/galileo/experiments.py</li>
<li>tests/test_experiments.py</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>fernando.correia@galil...</td><td>chore: update ruff tar...</td><td>April 07, 2026</td></tr>
<tr><td>thiago.bomfin@galileo.ai</td><td>fix: regression issues...</td><td>April 05, 2026</td></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/rungalileo/galileo-python/537?tool=ast>(Baz)</a>.

### CODE REVIEW

File: src/galileo/experiments.py [hunk 1]
```diff
@@ -214,7 +223,7 @@ def logged_process_func(row: DatasetRecord) -> Callable:
                 galileo_context.reset_trace_context()
                 if getsizeof(results) > MAX_REQUEST_SIZE_BYTES or len(results) >= MAX_INGEST_BATCH_SIZE:
                     _logger.info("Flushing logger due to size limit")
-                    galileo_context.flush()
+                    galileo_context.flush(on_error=on_error)
                     results = []
         # For dataset object, paginate through content
```
[the remark this answers is not in the document]
Comment (thiagobomfin-galileo) [reply src/galileo/experiments.py]: The duplication is 4 lines, it's immediately readable, and the alternative either introduces awkward mutation semantics or requires a larger restructuring. The on_error addition didn't create a meaningful new abstraction opportunity here.

**Review by fernandocorreia-galileo — CHANGES_REQUESTED** [review]
- **[Accept]**: baz-reviewer on `experiments.py:191` (line 190): `on_error(e)` in `Experiments.run()` is not guarded with try/except, unlike the same pattern in `decorator.py`.
  - **Analysis**: Valid. In `decorator.py`, the callback is wrapped in try/except to prevent callback exceptions from escaping. In `Experiments.run()`, no such guard exists. If the user's `on_error` raises, the exception escapes `run()` — inconsistent with the resilient pattern established in `flush()`.
  - **Recommended action**: Wrap `on_error(e)` in `Experiments.run()` in a try/except, matching the pattern from `decorator.py:1121-1124`.

- **[Accept]**: baz-reviewer on `experiments.py:336`: Docstring says `on_error` handles flush errors, but the implementation also catches creation failures in `Experiments.run()`.
  - **Analysis**: Valid. The docstring on `run_experiment()` says "Optional callback invoked with the exception when a flush error occurs" but `on_error` is also invoked when `Experiments.create()` fails. This is misleading.
  - **Recommended action**: Update the docstring to: "Optional callback invoked with the exception when a flush or experiment creation error occurs. If None, errors are raised (for creation) or logged as warnings (for flush). Defaults to None."

- **[Accept]**: baz-reviewer on `experiments.py:430`: `result["experiment"]` is accessed after `Experiments().run()` which can now return `{}` when `on_error` is provided.
  - **Analysis**: **This is a real bug.** When `on_error` is provided and `Experiments.create()` fails, `run()` returns `{}`. Then at line 430, `if experiment_tags is not None:` proceeds to `result["experiment"]` which raises `KeyError`. This crashes the caller despite the intent of `on_error` being to avoid crashes.
  - **Recommended action**: Add an early return after the `run()` call: `if not result: return result` before the `experiment_tags` block.

- **[Critical]**: `KeyError` crash when `on_error` is provided and `Experiments.create()` fails in the prompt-template flow.
  - File: `src/galileo/experiments.py` (lines 430-431)
  - When `Experiments().run()` returns `{}` (because `on_error` handled the creation failure), `result["experiment"]` at line 431 raises `KeyError`, defeating the purpose of `on_error`.
  - Suggestion: Add `if not result: return result` after line 428, before the `experiment_tags` block.

- **[Critical]**: All 4 new decorator tests are **failing** on the current branch.
  - Files: `tests/test_decorator.py` (lines 1555-1672)
  - Root cause: The tests mock `ingest_traces` to raise, expecting the exception to reach `GalileoDecorator.flush()`. But `GalileoLogger.flush()` is decorated with `@warn_catch_exception(exceptions=(Exception,))` (at `logger/logger.py:1735`), which catches and swallows the exception before it ever reaches the decorator's try/except. The `on_error` callback is never invoked.
  - This means the `on_error` mechanism in `GalileoDecorator.flush()` is **dead code for flush errors** — the logger's own error handling catches everything first. The `on_error` callback can only be triggered by errors in `get_logger_instance()` (initialization), not in the actual flush.
  - Suggestion: This is the core architectural problem described in SC-56022. To make `on_error` work for flush errors, either: (a) pass `on_error` down into `GalileoLogger.flush()` and let it bypass `@warn_catch_exception`, or (b) add a mechanism to `GalileoLogger` to propagate errors back to callers (e.g., a `raise_on_error` flag that disables the decorator). Option (a) is more targeted; option (b) aligns with the original story's implementation plan.

- **[Major]**: `on_error(e)` in `Experiments.run()` (line 190) is not guarded with try/except.
  - File: `src/galileo/experiments.py` (line 190)
  - Inconsistent with the guarded pattern in `decorator.py:1121-1124`. If the user's callback raises, the exception escapes unhandled.
  - Suggestion: Wrap in try/except matching the decorator pattern:
    ```python
    try:
        on_error(e)
    except Exception as cb_exc:
        _logger.warning(f"Experiment.run on_error callback raised: {cb_exc}")
    ```

- **[Major]**: Docstring for `on_error` in `run_experiment()` is inaccurate.
  - File: `src/galileo/experiments.py` (lines 333-335)
  - Says "invoked with the exception when a flush error occurs" but the callback is also invoked for experiment creation failures.
  - Suggestion: Update to reflect both flush and creation error handling.

- **[Minor]**: 1 experiment test (`test_run_experiment_on_error_passed_to_flush_in_function_flow`) is also failing due to insufficient mocking.
  - File: `tests/test_experiments.py`
  - The test patches `galileo_context.flush` but the test setup doesn't fully mock the HTTP routes needed by the function flow, causing `AllMockedAssertionError`.

File: src/galileo/experiments.py [hunk 2]
```diff
@@ -413,6 +419,9 @@ def run_experiment(
         project_obj, dataset_obj, experiment_name, prompt_template, scorer_settings, prompt_settings
     )
```
Comment (fernandocorreia-galileo) [inline src/galileo/experiments.py]: `run_experiment()` never forwards `on_error` to `Experiments().run()` in the prompt-template flow**

In the prompt-template branch, the call is:
```python
result = Experiments().run(
    project_obj, dataset_obj, experiment_name, prompt_template, scorer_settings, prompt_settings
    # on_error is NOT passed here
)
```
The test `test_run_experiment_on_error_called_when_create_raises` expects `on_error` to be called when `Experiments.create` raises, but without forwarding, the exception propagates uncaught and the callback is never invoked.

**Review by fernandocorreia-galileo — CHANGES_REQUESTED** [review]

**Review by fernandocorreia-galileo — COMMENTED** [review]
> ⚠️ **This review was generated by an AI agent (Astra) and may contain mistakes. Please verify all suggestions independently.**

## Summary

Adds an optional `on_error` callback parameter to `GalileoLogger.flush()`, `GalileoDecorator.flush()`, `Experiments.run()`, `Experiments.run_with_function()`, and `run_experiment()`. When provided, flush/creation errors invoke the callback instead of raising or logging warnings. The PR removes the `@warn_catch_exception` decorator from `flush()` and `_flush_distributed`/`_flush_batch` in favor of inline error handling that supports the new callback.

**Verdict: needs_discussion** — The `on_error` callback behavior is inconsistent between the function flow and prompt-template flow in `run_experiment()`, and the design conflates resource management errors with telemetry errors in a way that may violate the project's error handling architecture.

## General Comments
- **🟠 major** (design): The `on_error` callback conflates two distinct error categories defined in CLAUDE.md: **resource management operations** (which should raise) and **telemetry/ingestion operations** (which should be resilient). `Experiments.run()` uses `on_error` to swallow `self.create()` failures — a resource management operation — which violates the project's error handling architecture. The PR description and the Shortcut ticket (SC-56022) are specifically about "flush errors causing 0 trace failures", which suggests `on_error` was intended for ingestion/flush errors only. Consider whether `on_error` should apply to creation failures at all, or whether it should be restricted to flush/ingestion operations only.
- **🟠 major** (bug): The `on_error` handling is inconsistent between the two `run_experiment()` flows:
- **Prompt-template flow**: `on_error` is forwarded to `Experiments().run()` which wraps `self.create()` in a try/except.
- **Function flow**: `Experiments().create()` at line 388 is called directly without `on_error` protection, so creation errors will propagate regardless of whether `on_error` is provided.

This means the same `on_error` parameter behaves differently depending on which code path is taken, which will surprise users.
## File Comments
- **src/galileo/experiments.py** — **🟡 minor** (documentation): Codecov reports 3 missing coverage lines in this file (81.25% patch coverage). Consider adding tests for the edge cases that aren't covered.
## Line Comments
- **src/galileo/experiments.py** (line 388) — **🟠 major** (bug): In the function flow, `Experiments().create()` is called WITHOUT `on_error` protection, but in the prompt-template flow (line 429), `Experiments().run()` wraps `self.create()` with `on_error`. This means if a user passes `on_error` and uses a `function`, creation failures will still raise unhandled — contradicting the docstring which says `on_error` handles "experiment creation or a flush" errors.

Either wrap this `create()` call in a try/except that invokes `on_error` (consistent with the prompt-template flow), or document that `on_error` only applies to flush errors in the function flow.
- **src/galileo/logger/logger.py** (lines 1761–1767) — **🟡 minor** (design): When `on_error` is provided, the original exception `e` is passed to the callback, but the callback's own exceptions are silently suppressed via `contextlib.suppress(Exception)` with no logging. This means if the user's `on_error` callback itself raises, the error is silently lost at this layer. The decorator layer above (`_on_flush_error`) does log callback failures, so in practice when called through the decorator this is fine. But if someone calls `GalileoLogger.flush(on_error=...)` directly, callback failures are invisible.

Consider adding a log statement inside the suppress, e.g.:
```python
try:
    on_error(e)
except Exception as cb_exc:
    self._logger.warning(f"on_error callback raised: {cb_exc}")
```
- **src/galileo/experiments.py** (lines 335–338) — **🟡 minor** (documentation): The updated docstring now correctly says "experiment creation or a flush" — good that this was updated from the earlier version. However, the behavior is actually asymmetric: in the function flow, `on_error` is only forwarded to `flush()` calls but NOT to the initial `Experiments().create()` call at line 388. The docstring promises more than the code delivers for the function flow.
- **src/galileo/experiments.py** (lines 433–434) — **🟢 praise** (bug): Good addition of the guard against `result` being empty or missing the `experiment` key. This correctly prevents a `KeyError` when `Experiments().run()` returns `{}` via the `on_error` path. This was flagged in review and addressed properly.
- **src/galileo/decorator.py** (lines 1116–1124) — **🟢 praise** (design): The `_on_flush_error` wrapper is a well-designed pattern — it provides a unified error handler that works both as the `on_error` callback for `GalileoLogger.flush()` and as the fallback handler for `get_logger_instance()` failures. The layered protection (debug log + callback invocation + callback-failure handling) is thorough.
- **src/galileo/decorator.py** (lines 1126–1131) — **🟡 minor** (design): The outer `try/except` catching exceptions from `get_logger_instance()` is a good defensive pattern — it handles the case where initialization itself fails before `flush()` is even called. However, since `GalileoLogger.flush()` already has its own error handling and calls `_on_flush_error`, there's now a path where `_on_flush_error` could be called twice for the same error: once inside `GalileoLogger.flush()` via the `on_error` parameter, and once in this outer except block if `GalileoLogger.flush()` somehow re-raises despite its own handling. In practice this is unlikely (the logger's flush catches all exceptions), so this is just defensive depth, not a bug.
- **tests/test_decorator.py** (lines 1554–1671) — **🟢 praise** (testing): Excellent test coverage for the flush `on_error` feature. The four tests cover: (1) callback invoked on error, (2) warning logged without callback, (3) callback exception swallowed, and (4) debug-not-warning log level when callback is present. Good use of the Given/When/Then convention.
- **src/galileo/logger/logger.py** (lines 1734–1735) — **🟡 minor** (design): The `@warn_catch_exception` decorator was removed from `flush()` and the error handling was inlined. This is a reasonable approach since the `on_error` callback parameter requires custom error routing that can't be expressed with the generic decorator. However, note that `async_flush()` (line 1771) still uses `@async_warn_catch_exception` and does NOT support `on_error`. If async callers ever need the same `on_error` pattern, this will need to be updated for consistency.

**Review by fernandocorreia-galileo — COMMENTED** [review]
> ⚠️ **This review was generated by an AI agent (Astra) and may contain mistakes. Please verify all suggestions independently.**

**Verdict: needs_discussion** — The `on_error` docstring on `run_experiment()` claims it handles creation errors, but it actually only handles flush errors in the function flow — this mismatch will mislead users and should be corrected before merge.

## General Comments
- **🟠 major** (documentation): The `on_error` callback only works for flush errors in the function flow (`run_experiment(..., function=fn, on_error=cb)`). It is silently ignored in the prompt-template flow. The docstring on `run_experiment()` incorrectly claims it handles "creation or a flush" errors — it never handles creation errors. This mismatch between documented and actual behavior will confuse SDK users. Recommend fixing the docstring and/or logging a warning when `on_error` is provided but unused.
- **🟡 minor** (design): `async_flush()` does not support the new `on_error` callback — it still uses `@async_warn_catch_exception(exceptions=(Exception,))`. Users of the async API don't get the new feature. This is fine as a follow-up, but worth noting for consistency.
- **🟡 minor** (design): Fernando's two review comments (about `Experiments.run()` missing `on_error` and `run_experiment()` not forwarding to the prompt flow) appear to still be open/unresolved. The final code intentionally chose NOT to add `on_error` to `Experiments.run()` (correct per error handling architecture), but the reviewer's comments should be addressed/resolved with an explanation of why the design changed.

File: src/galileo/experiments.py [hunk 3]
```diff
@@ -319,6 +321,10 @@ def run_experiment(
         Optional function to run with the experiment
     experiment_tags
         Optional dictionary of key-value pairs to tag the experiment with
+    on_error
+        Optional callback invoked with the exception when an error occurs during experiment
+        creation or a flush. If None, creation errors are raised and flush errors are logged
+        as warnings. Defaults to None.
```
[the remark this answers is not in the document]
Comment (fernandocorreia-galileo) [reply src/galileo/experiments.py]: @thiagobomfin-galileo update docstring to indicate that `on_error` handles `flush` errors.

**Review by fernandocorreia-galileo — COMMENTED** [review]
> ⚠️ **This review was generated by an AI agent (Astra) and may contain mistakes. Please verify all suggestions independently.**

**Verdict: approve** — Well-structured feature addition with proper error handling layering, comprehensive tests, and accurate documentation. Previous review comments have been addressed in the final commit.

## Follow-ups after merging
- `src/galileo/logger/logger.py:1769-1786`: Consider adding `on_error` support to `async_flush()` for API consistency, so async callers get the same callback capability as sync callers.

**Review by fernandocorreia-galileo — COMMENTED** [review]
> ⚠️ **This review was generated by an AI agent (Astra) and may contain mistakes. Please verify all suggestions independently.**

**Verdict: approve** — Well-designed opt-in error callback for flush operations that addresses a real customer issue (SC-56022). The error handling architecture is correct, tests cover the key behaviors, and backward compatibility is preserved.

## General Comments
- **🟡 minor** (documentation): The PR description says "Experiments.run() also uses on_error to handle creation failures without raising when a callback is supplied," but the final code intentionally does NOT do this — creation errors always propagate, and `on_error` only applies to flush. The `run_experiment()` docstring is correct; only the PR body is stale from an earlier iteration. Consider updating the PR description to match the final behavior before merging.

## Follow-ups after merging
- `src/galileo/utils/singleton.py:228-268`: `GalileoLoggerSingleton.flush()` and `flush_all()` call `logger.flush()` without forwarding any `on_error` callback. If these methods are used directly (e.g., from atexit handlers), flush failures will always be logged as warnings with no caller control. Consider adding `on_error` support here for consistency.
- `src/galileo/logger/logger.py:1770-1787`: `async_flush()` doesn't support `on_error` — already tracked as SC-61945.

File: src/galileo/logger/logger.py [hunk 4]
```diff
@@ -1731,24 +1731,46 @@ def conclude(
         return current_parent
 
     @nop_sync
-    @warn_catch_exception(exceptions=(Exception,))
-    def flush(self) -> list[LoggedTrace]:
+    def flush(self, on_error: Callable[[Exception], None] | None = None) -> list[LoggedTrace]:
```
Comment (fercor-cisco) [inline src/galileo/logger/logger.py:1734]: **🟡 minor** (design): The `@nop_sync` decorator returns `None` when logging is disabled, but the return type annotation says `list[LoggedTrace]`. When `on_error` is provided and logging is disabled, the caller gets `None` instead of `[]` with no callback invoked. This is pre-existing (the old `@warn_catch_exception` had the same issue), but worth noting since `on_error` users might now have stronger expectations about the return value.

```suggestion
    @nop_sync
    def flush(self, on_error: Callable[[Exception], None] | None = None) -> list[LoggedTrace] | None:
```

**Review by fernandocorreia-galileo — COMMENTED** [review]
> ⚠️ **This review was generated by an AI agent (Astra) and may contain mistakes. Please verify all suggestions independently.**

**Verdict: approve** — Well-designed solution to a real customer problem (SC-56022). The on_error callback is properly layered, backward compatible, and thoroughly tested through multiple review iterations.

## General Comments
- **🟡 minor** (design): The PR only adds `on_error` to the sync `flush()` path. The async `async_flush()` (line 1776 of logger.py) still uses `@async_warn_catch_exception` with no callback capability. This means async callers (e.g., any future async experiment runner) cannot benefit from this feature. This is tracked in SC-61945 — just confirming that's intentional for this PR's scope.

## Follow-ups after merging
- `src/galileo/logger/logger.py:1977-1983`: The `except RuntimeError` in `terminate()` is now dead code. Since `flush()` handles all exceptions internally (logging warning and returning []), it will never raise RuntimeError to this caller. The try/except block could be simplified to just `self.flush()`.
- `src/galileo/logger/logger.py:1775-1793`: `async_flush()` should get the same `on_error` callback support as `flush()` for API consistency. Tracked in SC-61945.

**Comment by fernandocorreia-galileo** [comment]
## Follow-up stories

- [sc-62014](https://app.shortcut.com/galileo/story/62014) — Clean up dead `RuntimeError` catch in `logger.terminate()`
- [sc-62015](https://app.shortcut.com/galileo/story/62015) — `@nop_sync` returns `None` instead of `[]` when logging is disabled

**Review by fernandocorreia-galileo — APPROVED** [review]
